### PR TITLE
HTTP Client: Introduce go-conntrack into http client provider

### LINKED ISSUE #35257: HTTP Client: Implement `go-conntrack` to track client connections
**What would you like to be added**:

Ideally, we'd like to be able to track client connection based on certain labels. Currently this happens for Enterprise and we need to implement it in the OSS repo as well.

We think of using [go-conntrack](https://github.com/mwitkow/go-conntrack) for this.

Needs to be implemented into https://github.com/grafana/grafana-plugin-sdk-go repo.

**Special notes**:

The implementation is needed both for `v7.5.x` and `v8.0.x` onwards.

### PR DESCRIPTION
**Pull request by dsotirakis**
**What this PR does / why we need it**:

We introduce `go-conntrack` to be able to track http connections names/types, so we can instrument services better and also have a better connections overview.

**Which issue(s) this PR fixes**:

Fixes #35257

**Special notes for your reviewer**:

This fix is for `v7.5.x` versions - another PR will (probably?) need to be raised to include changes in the `grafana-plugin-sdk-go` repo.

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -282,7 +282,7 @@ func (s *Server) buildServiceGraph(services []*registry.Descriptor) error {
 		s.cfg,
 		routing.NewRouteRegister(middleware.ProvideRouteOperationName, middleware.RequestMetrics(s.cfg)),
 		localcache.New(5*time.Minute, 10*time.Minute),
-		httpclientprovider.New(s.cfg),
+		httpclientprovider.New(s.cfg, "server"),
```
Comment (dsotirakis) [inline pkg/server/server.go:285]: Something better needed here

File: pkg/infra/httpclient/httpclientprovider/http_client_provider.go [hunk 2]
```diff
@@ -29,5 +33,23 @@ func New(cfg *setting.Cfg) httpclient.Provider {
...
+// newConntrackRoundTripper takes a http.DefaultTransport and adds the Conntrack Dialer
+// so we can instrument outbound connections
+func newConntrackRoundTripper(name string) *http.Transport {
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.DialContext = conntrack.NewDialContextFunc(
+		conntrack.DialWithTracing(),
+		conntrack.DialWithName(name),
+		conntrack.DialWithDialer(&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
```
Comment (marefr) [inline pkg/infra/httpclient/httpclientprovider/http_client_provider.go:51]: Should take the created transport and use that, i.e. `conntrack.DialWithDialer(transport.Dial...)`

File: pkg/infra/httpclient/httpclientprovider/http_client_provider.go [hunk 3]
```diff
@@ -29,5 +33,23 @@ func New(cfg *setting.Cfg) httpclient.Provider {
...
 	})
 }
+
+// newConntrackRoundTripper takes a http.DefaultTransport and adds the Conntrack Dialer
+// so we can instrument outbound connections
+func newConntrackRoundTripper(name string) *http.Transport {
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.DialContext = conntrack.NewDialContextFunc(
+		conntrack.DialWithTracing(),
+		conntrack.DialWithName(name),
```
Comment (marefr) [inline pkg/infra/httpclient/httpclientprovider/http_client_provider.go:48]: How are the name used? Should probably use something similar like this? https://github.com/grafana/grafana/blob/ce513e4637e2f8fbe56bb8a8c9077bf57a798ca2/pkg/infra/httpclient/httpclientprovider/datasource_metrics_middleware.go#L65

File: pkg/infra/httpclient/httpclientprovider/http_client_provider.go [hunk 4]
```diff
@@ -2,17 +2,21 @@ package httpclientprovider
...
 	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/setting"
+	"github.com/mwitkow/go-conntrack"
 )
 
 var newProviderFunc = sdkhttpclient.NewProvider
 
 // New creates a new HTTP client provider with pre-configured middlewares.
-func New(cfg *setting.Cfg) httpclient.Provider {
+func New(cfg *setting.Cfg, connName string) httpclient.Provider {
```
Comment (marefr) [inline pkg/infra/httpclient/httpclientprovider/http_client_provider.go:19]: This is created once when Grafana starts. Think you want separate name per data source? See other comment

**Review by marefr — CHANGES_REQUESTED** [review]
See comments

**Comment by marefr** [comment]
Might be smarter to start with a PR targeting v7.5.x to get it out for v7.5.8 since that code is completely different than this which uses the HTTP client provider.

**Comment by marefr** [comment]
@dsotirakis so now when v7.5.x is fixed I think this should be trivial to update?

File: pkg/infra/httpclient/httpclientprovider/http_client_provider.go [hunk 5]
```diff
@@ -29,5 +33,26 @@ func New(cfg *setting.Cfg) httpclient.Provider {
...
+			newConntrackRoundTripper(datasourceName, transport)
+		},
 	})
 }
+
+// newConntrackRoundTripper takes a http.DefaultTransport and adds the Conntrack Dialer
+// so we can instrument outbound connections
+func newConntrackRoundTripper(name string, transport *http.Transport) *http.Transport {
+	transport.DialContext = conntrack.NewDialContextFunc(
+		conntrack.DialWithTracing(),
```
Comment (marefr) [inline pkg/infra/httpclient/httpclientprovider/http_client_provider.go:50]: Let's skip this one for now. Still have no idea how to expose this data. Perhaps something to look into later together with adding support for conntrack to grafana server?

**Review by marefr — CHANGES_REQUESTED** [review]
Looks good, but have some suggested changes

File: pkg/infra/httpclient/httpclientprovider/http_client_provider.go [hunk 6]
```diff
@@ -29,5 +33,26 @@ func New(cfg *setting.Cfg) httpclient.Provider {
...
+		conntrack.DialWithName(name),
+		conntrack.DialWithDialer(&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}),
```
[the remark this answers is not in the document]
Comment (dsotirakis) [reply pkg/infra/httpclient/httpclientprovider/http_client_provider.go]: Nice one

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**